source/particles/particle_handler.cc

```diff
@@ -1348,7 +1348,12 @@ namespace Particles
               current_cell, out_particle->get_location(), *mapping);
           Tensor<1, spacedim> vertex_to_particle =
             out_particle->get_location() - current_cell->vertex(closest_vertex);
-          vertex_to_particle /= vertex_to_particle.norm();
+
+          // Only normalize if vector is larger than zero, the algorithm below
+          // works fine for zero vectors.
+          if (vertex_to_particle.norm() >
+              100. * std::numeric_limits<double>::epsilon())
```

Comment:
I'm late to this comment, but shouldn't we try with a relative tolerance here? So something like
```c++
            if (vertex_to_particle.norm() >
              100. * std::numeric_limits<double>::epsilon() *
              current_cell->diameter())
```
or, if necessary, something similar that avoids the `sqrt` in `diameter()` (called via `norm()`).

Reply:
We have `Tensor::norm_square()` which can do that. I second the request to use a relative tolerance, but in reality I would like you to think about two cases: The vector is nonzero or the vector is zero (within a tolerance). In the former case, you can normalize and use the normalized vector. In the latter case, you should not be using the resulting vector at all in the algorithm below and use the zero vector instead, rather than continuing to use a non-normalized vector. That will likely mean that you should extend the `if (vector is non-zero)` to some of the code below as appropriate.